lib/Runtime/Library/JavascriptObject.cpp

```diff
@@ -2060,4 +2060,195 @@ namespace Js
         }
         return returnValue;
     }
+
+    /*static*/
+    void JavascriptObject::CopyDataPropertiesHelper(Var source, RecyclableObject* to, PropertyId* excluded, uint32 excludedLength, ScriptContext* scriptContext)
+    {
+        // 1. Assert Type(to) is Object.
+        // 2. Assert Type(excluded) is List.
+        // 3. If from is undefined or null, let keys be a new empty List
+        // 4. Else,
+        //        a. Let from be ToObject(source).
+        //        b. Let keys be from.[[OwnpropertyKeys]]().
+        // 5. Repeat for each element nextKey of keys in List order,
+        //        a. Let found be false.
+        //        b. Repeat for each element e of excluded,
+        //            i. If e is not empty and SameValue(e, nextKey) is true, then
+        //                1. Set found to true.
+        //        c. If found is false, then
+        //            i. Let desc be from.[[GetOwnProperty]](nextKey).
+        //            ii. If desc is not undefined and desc.[[Enumerable]] is true, then
+        //                1. Let propValue be Get(from, nextKey).
+        //                2. Perform CreateDataProperty(to, nextKey, propValue).
+        // 6. Return target (target is returned in pointer).
+        Assert(JavascriptOperators::IsObject(to));
+        RecyclableObject* from = nullptr;
+        if (!JavascriptConversion::ToObject(source, scriptContext, &from))
+        {
+            if (JavascriptOperators::IsUndefinedOrNull(source))
+            {
+                return;
+            }
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_NeedObject, _u("Object.CopyDataPropertiesHelper"));
+        }
+
+#if ENABLE_COPYONACCESS_ARRAY
+        JavascriptLibrary::CheckAndConvertCopyOnAccessNativeIntArray<Var>(from);
+#endif
+        // if proxy, take slow path by calling [[OwnPropertyKeys]] on source
+        if (JavascriptProxy::Is(from))
+        {
+            CopyDataPropertiesForProxyObjects(from, to, excluded, excludedLength, scriptContext);
+        }
+        // else use enumerator to extract keys from source
+        else
+        {
+            CopyDataPropertiesForGenericObjects(from, to, excluded, excludedLength, scriptContext);
+        }
+    }
+
+    /*static*/
+    void JavascriptObject::CopyDataPropertiesForGenericObjects(RecyclableObject* from, RecyclableObject* to, PropertyId* excluded, uint32 excludedLength, ScriptContext* scriptContext)
+    {
+        // Steps 4 and 5 of CopyDataPropertiesHelper
+
+        EnumeratorCache* cache = scriptContext->GetLibrary()->GetObjectAssignCache(from->GetType());
+        JavascriptStaticEnumerator enumerator;
+        if (!from->GetEnumerator(&enumerator, EnumeratorFlags::SnapShotSemantics | EnumeratorFlags::EnumSymbols | EnumeratorFlags::UseCache, scriptContext, cache))
```

Comment:
Will this only get own properties, or all enumerable properties including those from the prototype chain?

Reply:
Also, looks like you include symbols here, while the spec looks like it refers to `OwnPropertyKeys` which I thought excluded symbols?

Reply:
So, this line comes directly from the Object.assign() implementation. I believe it only gets own properties, and symbols are included as potential property names. I will add some unit tests to make sure of this.

Reply:
Ok, I've updated new code addressing everything you've mentioned so far. Let me know if I missed anything.

Reply:
`OwnPropertyKeys` returns symbol-keyed properties, right? `Object.getOwnPropertyNames` is the one which skips symbols.

Reply:
Yeah, from what I've seen, Object.assign() (with the same spec) has this behavior.

Reply:
Ok, I just compared with v8 and it does look like symbols are included, but not things in the prototype chain. Does the enumerator here only look at own properties, and not go up the prototype chain?

Reply:
Yeah, the enumerator only looks at own properties.